README.md
# G5 Authenticatable

G5 Authenticatable provides a default authentication solution for G5
Rails applications. This gem configures and mounts
[Devise](https://github.com/plataformatec/devise) with a default User
model, using [OmniAuth](https://github.com/intridea/omniauth) to authenticate
to the G5 Auth server. Helpers are also provided to secure your API methods.

If you are already using Devise with your own model, this is not the
library you are looking for. Consider using the
[devise_g5_authenticatable](https://github.com/G5/devise_g5_authenticatable)
extension directly.

If you have a stand-alone service without a UI, you may not need Devise at
all. Consider using the
[g5_authenticatable_api](https://github.com/G5/g5_authenticatable_api)
library in isolation.

## Current Version

0.7.4

## Requirements

* [rails](https://github.com/rails/rails) >= 3.2

## Installation

1. Set the environment variable `DEVISE_SECRET_KEY` to the value
   generated by:

   ```console
   rake secret
   ```

2. Add this line to your application's Gemfile:

   ```ruby
   gem 'g5_authenticatable'
   ```

3. And then execute:

   ```console
   bundle
   ```

4. Run the generator:

   ```console
   rails g g5_authenticatable:install
   ```

   This creates a migration for the new g5_authenticatable_users
   table, and mounts the engine at `/g5_auth`.

5. Update your database:

   ```console
   rake db:migrate db:test:prepare
   ```

## Configuration

### Root route

Devise requires you to define a root route in your application's
`config/routes.rb`. For example:

```ruby
root :to => 'home#index'
```

### Registering your OAuth application

1. Visit the auth server admin console and login:
  * For development, visit https://dev-auth.g5search.com/admin
  * For production, visit https://auth.g5search.com/admin
2. Click "New Application"
3. Enter a name that recognizably identifies your application.
4. Enter the redirect URI where the auth server should redirect
   after the user successfully authenticates. It will be
   of the form `http://<your_host>/g5_auth/users/auth/g5/callback`.

   For non-production environments, this redirect URI does not have to
   be publicly accessible, but it must be accessible from the browser
   where you will be testing (so using something like
   http://localhost:3000/g5_auth/users/auth/g5/callback is fine if your browser
   and client application server are both local).

   If you are using the production G5 Auth server, the redirect URI **MUST**
   use HTTPS.
5. For a trusted G5 application, check the "Auto-authorize?" checkbox. This
   skips the OAuth authorization step where the user is prompted to explicitly
   authorize the client application to access the user's data.
6. Click "Submit" to obtain the client application's credentials.

### Environment variables

Once you have your OAuth 2.0 credentials, you'll need to set the following
environment variables for your client application:

* `G5_AUTH_CLIENT_ID` - the OAuth 2.0 application ID from the auth server
* `G5_AUTH_CLIENT_SECRET` - the OAuth 2.0 application secret from the auth server
* `G5_AUTH_REDIRECT_URI` - the OAuth 2.0 redirect URI registered with the auth server
* `G5_AUTH_ENDPOINT` - the endpoint URL (without any path info) for the G5 auth server.
  Generally, this will be set to either `https://dev-auth.g5search.com` or
  `https://auth.g5search.com` (the default).

If you need to make server-to-server API calls that are not associated with an
end user, you can also set up a default user's credentials with:

* `G5_AUTH_USERNAME` - the G5 auth server user name
* `G5_AUTH_PASSWORD` - the G5 auth server user's password

### Token validation

By default, G5 Authenticatable only validates access tokens on incoming API
requests. If you are relying on session-based authentication instead, it is
possible for the local application's session to remain active after the
global access token has been revoked.

If you want to guarantee that the local session is destroyed when the access
token is revoked, add the following to
`config/initializers/g5_authenticatable.rb`:

```ruby
G5Authenticatable.strict_token_validation = true
```

## Usage

### Controller filters and helpers

G5 Authenticatable installs all of the usual devise controllers and helpers.
To set up a controller that requires authentication, use this before_filter:

```ruby
before_filter :authenticate_user!
```

To verify if a user is signed in, use the following helper:

```ruby
user_signed_in?
```

To access the model instance for the currently signed-in user:

```ruby
current_user
```

To access scoped session storage:

```ruby
user_session
```

### Route helpers

There are several generic helper methods for session and omniauth
URLs. To sign in:

```ruby
new_session_path(:user)
```

To sign out:

```ruby
destroy_session_path(:user)
```

There are also generic helpers for the OmniAuth paths, though you
are unlikely to ever use these directly. The OmniAuth entry point
is mounted at:

```ruby
g5_authorize_path(:user)
```

And the OmniAuth callback is:

```ruby
g5_callback_path(:user)
```

You may be more familiar with Devise's generated scoped URL helpers.
These are still available, but are isolated to the engine's namespace.
For example:

```ruby
g5_authenticatable.new_user_session_path
g5_authenticatable.destroy_user_session_path
```

### Access token ###

When a user authenticates, their OAuth access token will be stored on
the local user:

```ruby
current_user.g5_access_token
```

This is to support server-to-server API calls with G5 services that are
protected by OAuth.

### Securing an API ###

#### Grape ####

If you include `G5AuthenticatableApi::Helpers::Grape`, you can use the
`authenticate_user!` method to protect your API actions:

```ruby
class MyApi < Grape::API
  helpers G5AuthenticatableApi::GrapeHelpers

  before { authenticate_user! }

  # ...
end
```

#### Rails ####

You can secure API actions that respond to json using the `authenticate_api_user!`
method:

```ruby
class MyResourcesController < ApplicationController
  respond_to :json

  before_filter :authenticate_api_user!

  def get
    @resource = MyResource.find(params[:id])
    respond_with(@resource)
  end
end
```

#### Secure Clients ####

If you have an ember application, no client-side changes are necessary to use a
secure API method, as long as the action that serves your ember app requires
users to authenticate with G5 via devise.

Any non-browser clients must use token-based authentication. In contexts where
a valid OAuth 2.0 access token is not already available, you may request a new
token from the G5 Auth server using
[g5_authentication_client](https://github.com/G5/g5_authentication_client).
Clients may pass the token to secure API actions either in the HTTP
Authorization header, or in a request parameter named `access_token`.

For more details, see the documentation for
[g5_authenticatable_api](https://github.com/G5/g5_authenticatable_api).

### Authorization ###

#### User Roles ####

User roles are defined on the auth server and automatically populated in the local
model layer when a user logs in:

```ruby
current_user.roles
# => #<ActiveRecord::Associations::CollectionProxy [#<G5Authenticatable::Role id: 1, name: "viewer", ...>]>
```

We use [rolify](https://github.com/RolifyCommunity/rolify) for role management,
which provides an interface for querying role assignments:

```ruby
current_user.has_role?(:editor)
```

G5 currently supports four different roles: `:super_admin`, `:admin`,
`:editor`, and `:viewer` (the default role).

Two convenience methods have been added to the `G5Authenticatable::User`:

* `user.clients` will return a list of clients that the user has any access to. Will return all clients if the user has a global role
* `user.client_roles` will return a list of roles that relate directly to a client

#### Policies and Scopes ####

G5 Authenticatable uses [pundit](https://github.com/elabs/pundit) to encapsulate
the authorization logic in policy objects. The pundit documentation contains a much
more thorough explanation of how to define and use policies, but a quick overview
is provided here.

The G5 Authenticatable generator created an `app/policies/application_policy.rb`
file in your project:

```ruby
class ApplicationPolicy < G5Authenticatable::BasePolicy
  class Scope < BaseScope
  end
end
```

The `G5Authenticatable::BasePolicy` and `G5Authenticatable::BasePolicy::BaseScope`
implement a set of default authorization rules that essentially forbids access
to all actions on all model instances unless the user has the `:super_admin`
role. It also provides a set of helper methods for checking user roles:
`super_admin?`, `admin?`, `editor?`, or `viewer?`.

In order to implement a custom policy for one of your application's models, you
can create a new policy in the `app/policies` directory. For instance, if you
have a `Widget` model, and you want to also grant permissions to update that
model to users with `:admin` or `:editor` roles:

```ruby
# app/policies/widget_policy.rb

class WidgetPolicy < ApplicationPolicy
  def update?
    super_admin? || admin? || editor?
  end
end
```

You also have access to the record being authorized, and can define rules based
on that. For instance, if you want to restrict `Widget` deletion based on both
user role and some flag on the `Widget` instance to be deleted:

```ruby
# app/policies/widget_policy.rb

class WidgetPolicy < ApplicationPolicy
  def destroy?
    (super_admin? || admin?) && !record.published?
  end
end
```

In order to implement data-level authorization, you can define a custom scope
within your policy. The scope `resolve` method should only return the records
to which the current user has access. You have access to the current `user` and
also the `scope` object (defaults to the record class). For instance, if a user
must be the owner of a widget in order to access it, but super admins are allowed
to access all widgets:

```ruby
# app/policies/widget_policy.rb

class WidgetPolicy < ApplicationPolicy
  class Scope < Scope
    def resolve
      if user.has_role?(:super_admin?)
        scope.all
      else
        scope.where(owner_id: user.id)
      end
    end
  end
end
```

If you want to apply the same authorization logic across all models in your
application, you can define them in `ApplicationPolicy` or
`ApplicationPolicy::Scope`.

#### Using Policies in Controllers ####

You can use the `authorize` method in your controller to ensure that the
current user has access to the current action on a particular model instance.
For instance, in your controller:

```ruby
class WidgetsController < ApplicationController
  # ...

  def update
    @widget = Widget.find(params[:id])
    authorize(@widget)

    if @widget.update_attributes(widget_params)
      flash[:notice] = "Widget successfully updated."
    end

    respond_with(@widget)
  end
end
```

In this example, the `authorize` method will automatically lookup the correct policy
and populate it with the `current_user` and the model argument, and then call the
`update?` method on it, based on the current action name.

You can directly access the policy scope to look up the records to which the
current user has access by using the `policy_scope` method:

```ruby
class WidgetsController < ApplicationController
  # ...

  def index
    @widgets = policy_scope(Widget)
  end
```

In this example, the `policy_scope` method will automatically look up the
`WidgetPolicy::Scope`, populate it with the current user and `Widget` record
class, and call the scope `resolve` method to retrieve the results.

#### Using Policies in Views ####

Sometimes, you want to be able to access authorization logic inside Rails views.
For instance, you may want to hide the link to edit a record if the user does
not have access to update that record. In these cases, you can use the `policy`
method to lookup the instance of the policy directly:

```erb
<% if policy(@widget).update? %>
  <%= link_to 'Edit', edit_widget_path(@widget) %>
<% end %>
```

You can also call the `policy_scope` method inside views.

If you are not rendering your views within your Rails application (for
instance, if you are using an Ember front-end), then you will have to build
your own API for querying policy information in order to access this
functionality directly in your view.

### Test Helpers ###

G5 Authenticatable currently only supports [rspec-rails](https://github.com/rspec/rspec-rails).
Helpers and shared contexts are provided for integration testing secure pages
and API methods.

#### Prerequisites ####

Because the test helpers are optional, bundler will not automatically install
these dependencies. You will have to add the following gems to your own Gemfile:

* [rspec-rails](https://github.com/rspec/rspec-rails)
* [factory_girl_rails](https://github.com/thoughtbot/factory_girl_rails)
* [webmock](https://github.com/bblimke/webmock)

#### Incompatibilities ####

There are [known issues](https://github.com/G5/g5_authenticatable/issues/27) using the
auth test helpers with [selenium-webdriver](http://docs.seleniumhq.org/projects/webdriver/)
and the Firefox browser. As such, selenium-webdriver is not officially supported by
the G5 Authenticatable library.

If you are using [capybara](https://github.com/jnicklas/capybara) to run your
integration tests, we highly recommend using
[poltergeist](https://github.com/teampoltergeist/poltergeist) for PhantomJS as
your javascript driver instead.

#### Installation ####

To automatically mix in helpers to your feature and request specs, include the
following line in your `spec/spec_helper.rb`:

```ruby
require 'g5_authenticatable/rspec'
```

Note that the example code assumes you are using Rspec 3 metadata syntax. If you are
using Rspec 2.x, you can either:

* Enable this syntax with the following config in your `spec/spec_helper.rb`:

  ```ruby
  RSpec.configure do |config|
    config.treat_symbols_as_metadata_keys_with_true_values = true
  end
  ```

* Or you can replace every instance of a bare symbol in the metadata with a full
  key-value pair. For example:

  ```ruby
  # This won't work out of the box with Rspec 2
  context 'my secure context', :auth do
    it 'can see my secrets'
  end

  # But this will...
  context 'my secure context', auth: true do
    it 'can see my secrets'
  end
  ```

#### Feature Specs ####

The easiest way to use g5_authenticatable in feature specs is through
the shared auth context. This context creates a user (available via
`let(:user)`) and then authenticates as that user. To use the shared
context, simply include `:auth` in the RSpec metadata for your example
or group:

```ruby
context 'my secure context', :auth do
  it 'can access some resource' do
    visit('the place')
    expect(page).to ...
  end
end
```

If you prefer, you can use the helper methods from
`G5Authenticatable::Test::FeatureHelpers` instead of relying on the shared
context. For example:

```ruby
describe 'my page' do
  context 'with valid user credentials' do
    let(:my_user) { create(:g5_authenticatable_user, email: 'my.email@test.host') }
    before { stub_g5_omniauth(my_user) }

    it 'should display the secure page' do
      visit('the page')
      expect(page).to ...
    end
  end

  context 'with invalid OAuth credentials' do
    before { stub_g5_invalid_credentials }

    it 'should display an error' do
      visit('the page')
      expect(page). to ...
    end
  end

  context 'when user has previously authenticated' do
    let(:my_user) { create(:g5_authenticatable_user, email: 'my.email@test.host') }
    before { visit_path_and_login_with('some other path', my_user) }

    it 'should display the thing I expect' do
      visit('the page')
      expect(page).to ...
    end
  end
end
```

#### Request Specs ####

You can test API methods that have been secured with g5_authenticatable by
using the auth request shared context. This context creates a user (available
via `let(:user)`) and then automatically authenticates as that user. To use the
shared context, simply tag your example group with the `:auth_request` RSpec
metadata:

```ruby
describe 'my secure API', :auth_request do
  it 'can access some resource' do
    get '/api/v1/resource'
    expect(response).to ...
  end
end
```

Alternatively, you may wish to use the helper methods from
`G5Authenticatable::Test::RequestHelpers` directly:

```ruby
describe 'my secure API' do
  context 'when user is authenticated' do
    before { login_user }

    it 'can access some resource' do
      get '/api/v1/resource'
      expect(response).to be_success
    end
  end

  context 'when there is no authenticated user' do
    before { logout_user }

    it 'cannot access the resource' do
      get '/api/v1/resource'
      expect(response.status).to eq(401)
    end
  end
end
```

#### Controller Specs ####

You can test controller specs that have been secured with g5_authenticatable
by using the controller spec shared context.  This context creates a user
available via `let(:user)` and then automatically authenticates as that user.
To use the shared context, tag your example group with the `:auth_controller`
Rspec metadata:

```ruby
describe 'my secure action' do
  context 'when the user is authenticated' do
    it 'can access some secure path' do
      get :my_action
      expect(response). to be_success
    end
  end

  context 'when there is no authenticated user', :auth_controller do
    it 'cannot access the secure path' do
      get :my_action
      expect(reponse).to be_redirect
    end
  end
end
```

#### Token Validation Helpers ####

If you tag your examples with auth metadata (e.g. `:auth`, `:auth_request` or
`:auth_controller`), then the shared context will automatically take care of
any stubs required to support strict token validation.

However, if you are using the auth test helper methods directly, and you have
enabled strict token validation, then you will need to use the methods in
`G5Authenticatable::Test::TokenValidationHelpers` to stub external calls to
validate the access token.

For example, in a feature spec, you could use the `stub_valid_access_token`
method like so:

```ruby
describe 'my page' do
  let(:user) { FactoryGirl.create(:g5_authenticatable_user) }

  before do
    stub_g5_omniauth(user)
    stub_valid_access_token(user.g5_access_token)
  end

  it 'should let me in'
end
```

As another example, in a request spec, you could stub a revoked access
token using the `stub_invalid_access_token` helper:

```ruby
describe 'my API call' do
  let(:user) { FactoryGirl.create(:g5_authenticatable_user) }

  before { login_user }

  context 'when token becomes invalid after login' do
    before { stub_invalid_access_token(user.g5_access_token) }

    it 'should return 401'
  end

  context 'when token remains valid after login' do
    before { stub_valid_access_token(user.g5_access_token) }

    it 'should return 200'
  end
end
```

The same token validation helpers are also available in controller
specs, or anywhere else that authentication logic may be invoked.

### Purging local user data

G5 Authenticatable automatically maintains user data locally via the
`G5Authenticatable::User` model. This local data can be purged
using the following rake task:

```console
$ rake g5_authenticatable:purge_users
```


CHANGELOG.md
## v0.7.5.beta (2016-01-15)

* Fixes bug that granted higher permissions when the resource for a scoped roles didn't exist.


## v0.7.4 (2015-12-09)

* Adds Impersonate (Assume) Devise Strategy and Concern to handle access to session stored values
* Handles the ability and rules to impersonate a user providing his unique ID
* Adds the UID from the session and request to be used by the Auth Workflow when present

## v0.7.3 (2015-11-30)
* Upgrade to devise_g5_authenticatable to fix regression bug https://github.com/G5/devise_g5_authenticatable/pull/23
* Upgrade to omniauth_g5 to fix regression bug https://github.com/G5/omniauth-g5/pull/10

## v0.7.2 (2015-06-23)

* Removed pinned g5_updatalbe version from 0.6.0 to > 0.6.0

## v0.7.1 (2015-06-08)

* Fixed bug when listing clients

## v0.7.0 (2015-06-04)

* Add convenience methods to `G5Authenticatable::User`, as well as a
  `G5Updatable::ClientPolicy` for authorizing access to clients based on roles
  ([#37](https://github.com/G5/g5_authenticatable/pull/37))

## v0.6.0 (2015-06-02)

* Add support for resource-scoped roles
  ([#35](https://github.com/G5/g5_authenticatable/pull/35))

## v0.5.1 (2015-06-01)

* Upgraded dependency on
  [devise_g5_authenticatable](https://github.com/devise_g5_authenticatable) to
  pick up compatibility fixes for devise v3.5.1
  ([#34](https://github.com/G5/g5_authenticatable/issues/34))
* Removed custom `G5Authenticatable::FailureApp`, as the fix to devise itself
  was released ([#25](https://github.com/G5/g5_authenticatable/issues/25))

## v0.5.0 (2015-05-21)

* Added user roles. Requires running `rails g g5_authenticatable:install` and
  `rake db:migrate`
  ([#33](https://github.com/G5/g5_authenticatable/pull/33))
* Added user attributes. Requires running `rails g g5_authenticatable:install`
  and `rake db:migrate`
  ([#32](https://github.com/G5/g5_authenticatable/pull/32))
* Updated documentation around test helper dependencies and incompatibilities
  ([#30](https://github.com/G5/g5_authenticatable/pull/30))

## v0.4.2 (2015-02-10)

* Override `Devise::FailureApp` with fix for compatibility with Rails 4.2
  ([#26](https://github.com/G5/g5_authenticatable/pull/26))

## v0.4.1 (2015-01-26)

* Fix test helpers when strict token validation is enabled during testing
  ([#24](https://github.com/G5/g5_authenticatable/pull/24))

## v0.4.0 (2015-01-20)

* Several fixes around sign-out, including accepting GET requests and
  enabling strict token validation
  ([#21](https://github.com/G5/g5_authenticatable/pull/21))
* Improved documentation around controller test helpers
  ([#22](https://github.com/G5/g5_authenticatable/pull/22))
* Added `g5_authenticatable:purge_users` rake task to purge local user data;
  primarily used for configuring demo/dev environments built on a production
  clone or DB dump
  ([#23](https://github.com/G5/g5_authenticatable/pull/23))

## v0.3.0 (2014-03-13)

* First open source release to [RubyGems](https://rubygems.org)

## v0.2.0 (2014-03-11)

* Controller test helpers

## v0.1.4 (2014-03-07)

* Update dependency on [g5_authenticatable_api](https://github.com/G5/g5_authenticatable_api)
  for bug fix to ignore password credential configuration during token validation.

## v0.1.3 (2014-03-06)

* Remove auth client defaults in favor of env variables

## v0.1.2 (2014-03-05)

* Set `G5_AUTH_USERNAME` and `G5_AUTH_PASSWORD` on auth client defaults

## v0.1.0 (2014-02-26)

* Update dependency on [g5_authenticatable_api](https://github.com/G5/g5_authenticatable_api)
  to include new Rails API helpers.
* Fix shared test helpers for client applications that do not mixin the FactoryGirl syntax methods
  in their RSpec config.

## v0.0.4 (2014-02-20)

* Integrate [g5_authenticatable_api](https://github.com/G5/g5_authenticatable_api)
  for securing API methods.

## v0.0.3 (2014-02-13)

* Test helpers and shared context for integration tests in client applications.

## v0.0.2 (2014-02-10)

* Bump version for [devise_g5_authenticatable](https://github.com/G5/devise_g5_authenticatable)
  to pick up PostgreSQL fix.

## v0.0.1 (2014-02-07)

* Initial release


config/initializers/devise.rb
require 'g5_authenticatable/strategies/token_authenticatable'

# Use this hook to configure devise mailer, warden hooks and so forth.
# Many of these configuration options can be set straight in your model.
Devise.setup do |config|
  # The secret key used by Devise. Devise uses this key to generate
  # random tokens. Changing this key will render invalid all existing
  # confirmation, reset password and unlock tokens in the database.
  config.secret_key = ENV['DEVISE_SECRET_KEY']

  # ==> Mailer Configuration
  # Configure the e-mail address which will be shown in Devise::Mailer,
  # note that it will be overwritten if you use your own mailer class
  # with default "from" parameter.
  config.mailer_sender = 'do-not-reply@g5search.com'

  # Configure the class responsible to send e-mails.
  # config.mailer = 'Devise::Mailer'

  # ==> ORM configuration
  # Load and configure the ORM. Supports :active_record (default) and
  # :mongoid (bson_ext recommended) by default. Other ORMs may be
  # available as additional gems.
  require 'devise/orm/active_record'

  # ==> Configuration for any authentication mechanism
  # Configure which keys are used when authenticating a user. The default is
  # just :email. You can configure it to use [:username, :subdomain], so for
  # authenticating a user, both parameters are required. Remember that those
  # parameters are used only when authenticating and not when retrieving from
  # session. If you need permissions, you should implement that in a before filter.
  # You can also supply a hash where the value is a boolean determining whether
  # or not authentication should be aborted when the value is not present.
  # config.authentication_keys = [ :email ]

  # Configure parameters from the request object used for authentication. Each entry
  # given should be a request method and it will automatically be passed to the
  # find_for_authentication method and considered in your model lookup. For instance,
  # if you set :request_keys to [:subdomain], :subdomain will be used on authentication.
  # The same considerations mentioned for authentication_keys also apply to request_keys.
  # config.request_keys = []

  # Configure which authentication keys should be case-insensitive.
  # These keys will be downcased upon creating or modifying a user and when used
  # to authenticate or find a user. Default is :email.
  config.case_insensitive_keys = [ :email ]

  # Configure which authentication keys should have whitespace stripped.
  # These keys will have whitespace before and after removed upon creating or
  # modifying a user and when used to authenticate or find a user. Default is :email.
  config.strip_whitespace_keys = [ :email ]

  # Tell if authentication through request.params is enabled. True by default.
  # It can be set to an array that will enable params authentication only for the
  # given strategies, for example, `config.params_authenticatable = [:database]` will
  # enable it only for database (email + password) authentication.
  # config.params_authenticatable = true

  # Tell if authentication through HTTP Auth is enabled. False by default.
  # It can be set to an array that will enable http authentication only for the
  # given strategies, for example, `config.http_authenticatable = [:database]` will
  # enable it only for database authentication. The supported strategies are:
  # :database      = Support basic authentication with authentication key + password
  # config.http_authenticatable = false

  # If http headers should be returned for AJAX requests. True by default.
  # config.http_authenticatable_on_xhr = true

  # The realm used in Http Basic Authentication. 'Application' by default.
  # config.http_authentication_realm = 'Application'

  # It will change confirmation, password recovery and other workflows
  # to behave the same regardless if the e-mail provided was right or wrong.
  # Does not affect registerable.
  # config.paranoid = true

  # By default Devise will store the user in session. You can skip storage for
  # particular strategies by setting this option.
  # Notice that if you are skipping storage for all authentication paths, you
  # may want to disable generating routes to Devise's sessions controller by
  # passing :skip => :sessions to `devise_for` in your config/routes.rb
  config.skip_session_storage = [:http_auth]

  # By default, Devise cleans up the CSRF token on authentication to
  # avoid CSRF token fixation attacks. This means that, when using AJAX
  # requests for sign in and sign up, you need to get a new CSRF token
  # from the server. You can disable this option at your own risk.
  # config.clean_up_csrf_token_on_authentication = true

  # ==> Configuration for :database_authenticatable
  # For bcrypt, this is the cost for hashing the password and defaults to 10. If
  # using other encryptors, it sets how many times you want the password re-encrypted.
  #
  # Limiting the stretches to just one in testing will increase the performance of
  # your test suite dramatically. However, it is STRONGLY RECOMMENDED to not use
  # a value less than 10 in other environments.
  config.stretches = Rails.env.test? ? 1 : 10

  # Setup a pepper to generate the encrypted password.
  # config.pepper = '8d901e67527b2a2c9f4b6ba83bccb2d7103e3c5ae24f811d3e35834d6c279404dbdf31b45efaaba8aa304f0f9441fcf1dc243c641a182a99460b2651da82a483'

  # ==> Configuration for :confirmable
  # A period that the user is allowed to access the website even without
  # confirming his account. For instance, if set to 2.days, the user will be
  # able to access the website for two days without confirming his account,
  # access will be blocked just in the third day. Default is 0.days, meaning
  # the user cannot access the website without confirming his account.
  # config.allow_unconfirmed_access_for = 2.days

  # A period that the user is allowed to confirm their account before their
  # token becomes invalid. For example, if set to 3.days, the user can confirm
  # their account within 3 days after the mail was sent, but on the fourth day
  # their account can't be confirmed with the token any more.
  # Default is nil, meaning there is no restriction on how long a user can take
  # before confirming their account.
  # config.confirm_within = 3.days

  # If true, requires any email changes to be confirmed (exactly the same way as
  # initial account confirmation) to be applied. Requires additional unconfirmed_email
  # db field (see migrations). Until confirmed new email is stored in
  # unconfirmed email column, and copied to email column on successful confirmation.
  config.reconfirmable = true

  # Defines which key will be used when confirming an account
  # config.confirmation_keys = [ :email ]

  # ==> Configuration for :rememberable
  # The time the user will be remembered without asking for credentials again.
  # config.remember_for = 2.weeks

  # If true, extends the user's remember period when remembered via cookie.
  # config.extend_remember_period = false

  # Options to be passed to the created cookie. For instance, you can set
  # :secure => true in order to force SSL only cookies.
  # config.rememberable_options = {}

  # ==> Configuration for :validatable
  # Range for password length. Default is 8..128.
  config.password_length = 8..128

  # Email regex used to validate email formats. It simply asserts that
  # one (and only one) @ exists in the given string. This is mainly
  # to give user feedback and not to assert the e-mail validity.
  # config.email_regexp = /\A[^@]+@[^@]+\z/

  # ==> Configuration for :timeoutable
  # The time you want to timeout the user session without activity. After this
  # time the user will be asked for credentials again. Default is 30 minutes.
  config.timeout_in = 1.hour

  # If true, expires auth token on session timeout.
  # config.expire_auth_token_on_timeout = false

  # ==> Configuration for :lockable
  # Defines which strategy will be used to lock an account.
  # :failed_attempts = Locks an account after a number of failed attempts to sign in.
  # :none            = No lock strategy. You should handle locking by yourself.
  # config.lock_strategy = :failed_attempts

  # Defines which key will be used when locking and unlocking an account
  # config.unlock_keys = [ :email ]

  # Defines which strategy will be used to unlock an account.
  # :email = Sends an unlock link to the user email
  # :time  = Re-enables login after a certain amount of time (see :unlock_in below)
  # :both  = Enables both strategies
  # :none  = No unlock strategy. You should handle unlocking by yourself.
  # config.unlock_strategy = :both

  # Number of authentication tries before locking an account if lock_strategy
  # is failed attempts.
  # config.maximum_attempts = 20

  # Time interval to unlock the account if :time is enabled as unlock_strategy.
  # config.unlock_in = 1.hour

  # Warn on the last attempt before the account is locked.
  # config.last_attempt_warning = false

  # ==> Configuration for :recoverable
  #
  # Defines which key will be used when recovering the password for an account
  # config.reset_password_keys = [ :email ]

  # Time interval you can reset your password with a reset password key.
  # Don't put a too small interval or your users won't have the time to
  # change their passwords.
  config.reset_password_within = 6.hours

  # ==> Configuration for :encryptable
  # Allow you to use another encryption algorithm besides bcrypt (default). You can use
  # :sha1, :sha512 or encryptors from others authentication tools as :clearance_sha1,
  # :authlogic_sha512 (then you should set stretches above to 20 for default behavior)
  # and :restful_authentication_sha1 (then you should set stretches to 10, and copy
  # REST_AUTH_SITE_KEY to pepper).
  #
  # Require the `devise-encryptable` gem when using anything other than bcrypt
  # config.encryptor = :sha512

  # ==> Scopes configuration
  # Turn scoped views on. Before rendering "sessions/new", it will first check for
  # "users/sessions/new". It's turned off by default because it's slower if you
  # are using only default views.
  # config.scoped_views = false

  # Configure the default scope given to Warden. By default it's the first
  # devise role declared in your routes (usually :user).
  # config.default_scope = :user

  # Set this configuration to false if you want /users/sign_out to sign out
  # only the current scope. By default, Devise signs out all scopes.
  # config.sign_out_all_scopes = true

  # ==> Navigation configuration
  # Lists the formats that should be treated as navigational. Formats like
  # :html, should redirect to the sign in page when the user does not have
  # access, but formats like :xml or :json, should return 401.
  #
  # If you have any extra navigational formats, like :iphone or :mobile, you
  # should add them to the navigational formats lists.
  #
  # The "*/*" below is required to match Internet Explorer requests.
  # config.navigational_formats = ['*/*', :html]

  # The default HTTP method used to sign out a resource. Default is :delete.
  config.sign_out_via = :get

  # ==> OmniAuth
  # Add a new OmniAuth provider. Check the wiki for more information on setting
  # up on your models and hooks.
  # config.omniauth :github, 'APP_ID', 'APP_SECRET', :scope => 'user,public_repo'
  config.omniauth :g5, ENV['G5_AUTH_CLIENT_ID'], ENV['G5_AUTH_CLIENT_SECRET'],
                       client_options: {site: ENV['G5_AUTH_ENDPOINT']}

  # ==> Warden configuration
  # If you want to use other strategies, that are not supported by Devise, or
  # change the failure app, you can configure them inside the config.warden block.
  #
  config.warden do |manager|
    manager.failure_app = G5Authenticatable::FailureApp
  end


  # ==> Mountable engine configurations
  # When using Devise inside an engine, let's call it `MyEngine`, and this engine
  # is mountable, there are some extra configurations to be taken into account.
  # The following options are available, assuming the engine is mounted as:
  #
  #     mount MyEngine, at: '/my_engine'
  #
  # The router that invoked `devise_for`, in the example above, would be:
  # config.router_name = :my_engine
  config.router_name = :g5_authenticatable

  # When using omniauth, Devise cannot automatically set Omniauth path,
  # so you need to do it manually. For the users scope, it would be:
  config.omniauth_path_prefix = '/g5_auth/users/auth'
end
